File: src/main/java/core/procedures/strings/StringRef.java
package core.procedures.strings;

import core.procedures.AFn;
import core.procedures.FnArgsBuilder;
import core.scm.SCMClass;

public final class StringRef extends AFn {

  public StringRef() {
    super(new FnArgsBuilder().minArgs(2).maxArgs(2)
                             .mandatoryArgsTypes(new Class[]{String.class, SCMClass.ExactNonNegativeInteger.class}));
  }

  @Override
  public boolean isPure() {
    return true;
  }

  @Override
  public String getName() {
    return "string-ref";
  }

  @Override
  public Object apply2(Object arg1, Object arg2) {
    String s = arg1.toString();
    Long pos = ((Number)arg2).longValue();
    if (pos >= s.length()) {
      throw new IndexOutOfBoundsException(String.format("%s: value out of range: %s", getName(), pos));
    }
    return s.charAt(pos.intValue());
  }
}

File: src/main/java/core/procedures/strings/StringRef.kt
package core.procedures.strings

import core.procedures.AFn
import core.procedures.Arity.Exactly
import core.scm.Type

class StringRef : AFn<Any?, Char>(name = "string-ref", isPure = true, arity = Exactly(2),
                                  mandatoryArgsTypes = arrayOf(CharSequence::class.java,
                                                               Type.ExactNonNegativeInteger::class.java)) {

    override operator fun invoke(arg1: Any?, arg2: Any?) = (arg2 as Number).toInt().let {
        arg1!!.toString().getOrElse(it) { throw IndexOutOfBoundsException("$name: value out of range: $it") }
    }
}
